src/ireiat/data_pipeline/assets/demand/faf5_tons.py

```diff
@@ -11,54 +11,155 @@
+            io_manager_key="custom_io_manager",
+            metadata={"format": "parquet", **INTERMEDIATE_DIRECTORY_ARGS},
+        ),
+        "faf5_water_demand": dagster.AssetOut(
+            io_manager_key="custom_io_manager",
+            metadata={"format": "parquet", **INTERMEDIATE_DIRECTORY_ARGS},
+        ),
     }
 )
 def faf5_demand_by_mode(
     context: dagster.AssetExecutionContext, faf5_demand_src: pd.DataFrame
-) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame, pd.DataFrame]:
+) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]:
     """
-    Multi-asset function to extract demand data for different transport modes (Truck, Rail, Water, Other/Unknown).
-    Currently, FAF containerizable demand is limited by commodity.
+    Multi-asset function to extract demand data for different transport modes (Truck, Rail, Water).
+    Rows with unknown modes are distributed proportionally based on known modes for each origin-destination pair.
     """
-
     # Filter for containerizable commodities
     is_containerizable = ~faf5_demand_src["sctg2"].isin(NON_CONTAINERIZABLE_COMMODITIES)
     faf_containerizable_pdf = faf5_demand_src.loc[is_containerizable]
 
-    # Initialize a list to store DataFrames for each mode
-    demand_dataframes = []
+    # Split into known modes (truck, rail, water) and unknown modes
+    known_modes_pdf = faf_containerizable_pdf[
+        faf_containerizable_pdf["dms_mode"].isin([FAFMode.TRUCK, FAFMode.RAIL, FAFMode.WATER])
+    ]
+    unknown_modes_pdf = faf_containerizable_pdf[
+        faf_containerizable_pdf["dms_mode"] == FAFMode.OTHER_AND_UNKNOWN
+    ]
+
+    # Aggregate by origin/destination for known modes
+    known_agg_pdf = known_modes_pdf.groupby(["dms_orig", "dms_dest", "dms_mode"], as_index=False)[
+        [FAF_TONS_TARGET_FIELD]
+    ].sum()
+
+    # Pivot by origin/destination to calculate percentages for each mode
+    mode_pivot_pdf = known_agg_pdf.pivot_table(
+        index=["dms_orig", "dms_dest"],
+        columns="dms_mode",
+        values=FAF_TONS_TARGET_FIELD,
+        fill_value=0,
+    )
+
+    # Add total tons and calculate percentages
+    mode_pivot_pdf["total_tons"] = mode_pivot_pdf.sum(axis=1)
+    mode_pivot_pdf["truck_pct"] = mode_pivot_pdf[FAFMode.TRUCK] / mode_pivot_pdf["total_tons"]
+    mode_pivot_pdf["rail_pct"] = mode_pivot_pdf[FAFMode.RAIL] / mode_pivot_pdf["total_tons"]
+    mode_pivot_pdf["water_pct"] = mode_pivot_pdf[FAFMode.WATER] / mode_pivot_pdf["total_tons"]
+
+    # Filter unknown origin/destination pairs that don't exist in known modes
+    valid_orig_dest_pairs = known_agg_pdf[["dms_orig", "dms_dest"]].drop_duplicates()
+
+    # Calculate unique origin-destination pairs in unknown modes before filtering
+    original_unknown_pairs = unknown_modes_pdf[["dms_orig", "dms_dest"]].drop_duplicates()
+
+    # Merge unknown modes with the known mode proportions, filter invalid pairs
+    unknown_modes_pdf = unknown_modes_pdf.merge(
```

Comment:
Part of me doesn't like mutating this dataframe (i.e. assigning it to itself). Generally like to avoid...

Could be easier to just do an inner merge here to a `common_modes_pdf` and then take the set difference of keys between the unknown_modes[["dms_orig","dms_dest"]] and valid_orig_dest_pairs...

Reply:
Updated the logic based on your suggestion